Optimizer bugfixes

paulstapor opened the pull request:
A lot of little bugfixes, interface to lsqnonlin, new options for optimization etc...

Most important features:
 * Yannik's gradient-free opimizers are new on this branch (therefore the Bobyqa and Cs and DHC-Code are new)
 * lsqnonlin will be on the master now
 * lsqnonlin examples are implemented
 * getMultiStarts is split up: optimization is moved to in the "performOptimization***" routines, getMultiStarts wraps around this
 * figure handles are checked for validity
 * non-linear constraints are implemented
 * options are switched to CustomDisplay
 * the RafMekErk example is added in the non-hierarchical version
 * parameters can be fixed for optimization and profiling

Fixing the following issues:
#144 , #143 , #142 , #131 , #124 , #118 , #112 , #108 , #107 , #100 

All examples should be running without trouble again.
To the reviewers (especially @BBallnus  and @LoosC ): Please check if your things still work as they should. I tested them, it looked good, but I can not really judge the outcome.

FFroehlich commented:
I am not able to review a pull request of this size in coming weeks.

paulstapor commented on @DRAMOptions/DRAMOptions.m:
```diff
@@ -188,28 +188,3 @@
-
-
-
-
```
All the changes in this file are due to switching the inheritance from
matlab.mixin.SetGet to matlab.mixin.CustomDisplay

paulstapor commented on @MALAOptions/MALAOptions.m:
```diff
@@ -156,28 +156,3 @@
-
-
-
-
```
Again, changes in this file are due to switching the inheritance from
matlab.mixin.SetGet to matlab.mixin.CustomDisplay

paulstapor commented on @PHSOptions/PHSOptions.m:
```diff
@@ -204,28 +204,3 @@
-
-
-
-
```
Again, changes in this file are due to switching the inheritance from
matlab.mixin.SetGet to matlab.mixin.CustomDisplay

paulstapor commented on @PTOptions/PTOptions.m:
```diff
@@ -212,15 +212,16 @@
+      function this = set.temperatureEta(this, value)
          if(value == floor(value) && value > 0)
             this.temperatureEta = lower(value);
          else
```
Again, changes in this file are due to switching the inheritance from
matlab.mixin.SetGet to matlab.mixin.CustomDisplay

paulstapor commented on @PestoOptions/PestoOptions.m:
```diff
@@ -199,14 +199,7 @@
+        profileOptimizationOptions = [];
         
         % Maximal relative decrease of ratio allowed for two adjacent
         % points in the profile (default = 0.10) if options.dJ = 0;
```
Changes here are due to Issue #131:
Now, Profile Optimization will use the same options as local optimization, unless specified differently. So getParameterProfiles checks for emptiness of this option.
Moreover, it's renamed to profileOptimizationOptions instead of profile"Re"optimizationOptions.

paulstapor commented on @PestoPlottingOptions/PestoPlottingOptions.m:
```diff
@@ -388,63 +388,63 @@
+        function this = set.hold_on(this, value)
             if islogical(value)
                 this.hold_on = value;
             else
```
mixin.SetGet swichted to CustomDisplay

paulstapor commented on @PestoSamplingOptions/PestoSamplingOptions.m:
```diff
@@ -384,28 +384,3 @@
-
-
-
-
```
Again, mixin.SetGet switched to CustomDisplay...

paulstapor commented on bobyqa/LICENCE.txt:
```diff
@@ -0,0 +1,18 @@
+
+Michael J. D. Powell <mjdp@cam.ac.uk>
+University of Cambridge
+Cambridge, UK.
```
Bobyqua added as optimizer

paulstapor commented on bobyqa/Makefile:
```diff
@@ -0,0 +1,31 @@
+update.o: update.f
+	$(FC) $(FFLAGS) -c update.f
+clean:
+	rm -f *.o main
```
bobyqa added as local optimizer

dweindl replied:
Are there some compilation/installation instructions somewhere? If it is repackaged in PESTO, it should come with a simple matlab-callable compilation script. The average windows user might not be familiar with make.

yannikschaelte replied:
now there is a compilation file, and the program gives directions what to do

paulstapor commented on bobyqa/bobyqa.m:
```diff
@@ -0,0 +1,110 @@
+varargout{2} = fval;
+varargout{3} = exitflag;
+varargout{4} = output;
+end
```
Bobyqa added as local optimizer

paulstapor commented on examples/conversion_reaction/mainConversionReaction.m:
```diff
@@ -235,7 +255,7 @@
 % multi-start optimization runs by getPropertyMultiStarts.
 
-optionsProperties = optionsPesto.copy();
+optionsProperties = optionsPesto;
```
Due to mixin.SetGet -> CustomDisplay

paulstapor commented on examples/conversion_reaction/mainConversionReaction.m:
```diff
@@ -210,7 +231,6 @@
 % PT specific options:
 optionsPesto.MCMC.samplingAlgorithm   = 'PT';
 optionsPesto.MCMC.PT.nTemps           = 1;
-optionsPesto.MCMC.PT.temperatureAdaptionScheme = 'Lacki15'; %'Vousden16'; 
```
Dus to Sampling Options changed

paulstapor commented on examples/enzymatic_catalysis/logLikelihoodEC.m:
```diff
@@ -122,7 +122,7 @@
+    odeOptions = odeset('RelTol', 1e-4, 'AbsTol', 1e-8);
     [~,X] = ode15s(@(t,x) f(exp(theta),x), t, [con0(:,iMeasure); fillX], odeOptions);
     y = h(X(:,1:nStates), exp(theta));
     Y(:, :) = yMeasured(iMeasure, :, :);
```
sufficient for simulation and makes things run faster...

paulstapor commented on examples/enzymatic_catalysis/mainEnzymaticCatalysis.m:
```diff
@@ -108,7 +108,6 @@
 optionsPesto.MCMC.PT.nTemps         = 6;
 optionsPesto.MCMC.PT.exponentT      = 6;    
 optionsPesto.MCMC.PT.regFactor      = 1e-8;
-optionsPesto.MCMC.PT.temperatureAdaptionScheme = 'Lacki15'; %'Vousden16';
```
Due to sampling options changed

paulstapor commented on examples/jakstat_signaling/logLikelihoodJakstatLsqnonlin.m:
```diff
@@ -0,0 +1,70 @@
+        varargout{3} = sol.llh;
+    end
+    
+end
```
New likelihood function for presenting lsqnonlin on the Jakstat example...

paulstapor commented on examples/jakstat_signaling/mainJakstatSignaling.m:
```diff
@@ -52,9 +51,9 @@
-    warning('There was a problem with the AMICI toolbox (available at https:// github.com/ICB-DCM/AMICI), which is needed to run this example file. The original error message was:');
+    warning('This PESTO example uses the AMICI toolbox (available at https://github.com/ICB-DCM/AMICI). Unfortunately, there was a problem with AMICI when trying to run this example file. Please check if AMICI is properly installed to run this example. The original error message was:');
     rethrow(ME);
 end
```
More detailed warning, if Amici is not installed, related to #13

FFroehlich commented on doc/config/Doxyfile.template(1):
```diff
@@ -0,0 +1,2484 @@
+############################################################################
```
do we really need this file? Isn't this simply a replicate of Doxyfile.template?

FFroehlich commented on examples/jakstat_signaling/simulate_jakstat_pesto.m:
```diff
@@ -0,0 +1,243 @@
+% simulate_jakstat_pesto.m is the matlab interface to the cvodes mex
```
I dont think this file makes sense if we do not ship compiled mex simulation routines.

FFroehlich commented on examples/jakstat_signaling/logLikelihoodJakstatLsqnonlin.m:
```diff
@@ -0,0 +1,70 @@
+    % Set residuals
+    res = (sol.y(:) - amiData.Y(:)) ./ sol.sigmay(:);
+    res_sigma = sqrt(log(2*pi*sol.sigmay(:).^2) - log(2*pi*10^(-5)^2));
+    res([1 48]) = 0;
```
Please use a nanIndex instead of these magic numbers here (as you did for the rafmekerk example)

dweindl replied:
I second that. The numbers in the rest of this file don't make things very easy to follow.

FFroehlich commented on getMultiStarts.m:
```diff
@@ -117,61 +117,68 @@
     case 'silent' % no output
         % Force fmincon to be silent.
-        if strcmp(options.localOptimizer, 'fmincon')
+        if (strcmp(options.localOptimizer, 'fmincon') || strcmp(options.localOptimizer, 'lsqnonlin'))
```
can we simply check `isfield(options.localOptimizerOptions,'Display')`? Should be more robust for further extensions

FFroehlich commented on getMultiStarts.m:
```diff
@@ -117,61 +117,68 @@
 
 %% Initialization
-if strcmp(options.localOptimizer, 'fmincon') || strcmp(options.localOptimizer, 'pswarm')
+if (strcmp(options.localOptimizer, 'fmincon') || strcmp(options.localOptimizer, 'pswarm') || strcmp(options.localOptimizer, 'lsqnonlin'))
```
again check if the respective field exists? Only make special cases if for a particular optimizer the options does a different thing (which hopefully does not happen)

FFroehlich commented on getMultiStarts.m:
```diff
@@ -209,7 +216,7 @@
     
     % Matlab parallel toolbox seems to have problems with our outfun...
-    if strcmp(options.localOptimizer, 'fmincon')
+    if (strcmp(options.localOptimizer, 'fmincon') || strcmp(options.localOptimizer, 'lsqnonlin'))
```
check with isfield

dweindl commented on bobyqa/funHandleWrap.m:
```diff
@@ -0,0 +1,15 @@
+function [ y ] = funHandleWrap(x, calfun)
+% Wrapper to compute objective function value. Currently used global
+% variable in bobyqa.m, the argument calfun is not used.
```
"Currently used global variable in bobyqa.m," ? Don't understand.

yannikschaelte replied:
"currently uses global variable in bobyqa.m": means that currently the objective function here comes from a global variable defined in bobyqa.m. This is not-optimal design, but it is difficult/impossible to refer to a non-globally defined matlab function from within mex files. I will change the doc.

dweindl commented on examples/rafmekerk_signaling/logLikelihoodRME.m:
```diff
@@ -0,0 +1,63 @@
+ function [varargout] = logLikelihoodRME(theta, amiData)
+% compute the negative loglikelihood and gradient for RafMekErk data
```
Please document input parameters and especially `varargout`

dweindl commented on examples/rafmekerk_signaling/logLikelihoodRMELsqnonlin.m:
```diff
@@ -0,0 +1,81 @@
+ function [varargout] = logLikelihoodRMELsqnonlin(theta, amiData)
+% compute the negative loglikelihood and gradient for RafMekErk data
+
```
Please document input parameters and especially `varargout`. State difference to /logLikelihoodRME.m

dweindl commented on examples/rafmekerk_signaling/logLikelihoodRMELsqnonlin.m:
```diff
@@ -0,0 +1,81 @@
+        nCondition = size(amiData, 2);
+
+        % Initialize output
+        res = zeros(112, 1);
```
Lot's of magic numbers in this function.

dweindl commented on examples/rafmekerk_signaling/mainRafMekErkSignaling.m:
```diff
@@ -0,0 +1,91 @@
+% Hab ein bisschen auskommentiert, was genau wichtig ist
```
:thinking:

dweindl commented on examples/rafmekerk_signaling/mainRafMekErkSignaling.m:
```diff
@@ -0,0 +1,91 @@
+parameters.max(4) = 6;
+parameters.max(13:20) = 8;
+
+parameters.number =  n_theta + n_sigma;
```
why not move this 10 lines upwards and use it instead of magic 28?

dweindl commented on getFullStructs.m:
```diff
@@ -0,0 +1,108 @@
+function [parameters,properties] = getFullStructs(parameters,reducedParameters,properties,reducedProperties,options)
```
Hard to deduce what this function does from the name alone.

paulstapor replied:
General comment on this function:

This function is not used in the current code version. Jan started working on it in #121 to restructure the way how fixed parameters are currently handled.
However, I think this PR fixes enough issues already and we  should not include these files right now, even if we agreed on using them later. I suggest leaving the current implementation for fixed parameters for this PR.
Any objections to this idea?

dweindl commented on getFullStructs.m:
```diff
@@ -0,0 +1,108 @@
+        % Default values or results stored in parameters.P
+        for iPar = 1:parameters.number
+            flag_asign_default = true;
+            if isfield(parameters,'P')
```
Move before loop, does not depend on iPar

dweindl commented on getFullStructs.m:
```diff
@@ -0,0 +1,108 @@
+    if isfield(reducedParameters, 'P')
+        % Default values or results stored in parameters.P
+        for iPar = 1:parameters.number
+            flag_asign_default = true;
```
assign, +s

dweindl commented on getFullStructs.m:
```diff
@@ -0,0 +1,108 @@
+        for iPar = 1:parameters.number
+            flag_asign_default = true;
+            if isfield(parameters,'P')
+                if length(parameters.P) >= iPar
```
Can this be false? Would be an error worth mentioning then, or?

dweindl commented on getFullStructs.m:
```diff
@@ -0,0 +1,108 @@
+        properties.MS.par(options.fixedParameters, :) = options.fixedParameterValues;
+    end
+
+    % Assignment of profile calculation results
```
Feel free to split into individual functions (same file), that makes dependencies much clearer

dweindl commented on getParameterConfidenceIntervals.m:
```diff
@@ -62,53 +62,75 @@
+end
 
 % Loop: alpha levels
 for k = 1:length(alpha)
```
Who can follow 50 line loop bodies with 5 levels of indentation? I can't....

dweindl commented on getPropertyConfidenceIntervals.m:
```diff
@@ -47,52 +47,54 @@
-                    else
-                        properties.CI.PL(i,2,k) = inf;
-                    end
+            if ~isempty(properties.P(iProp).prop)
```
Looks all pretty redundant with getParameterConfidenceIntervals.m, can't some code be reused there?

dweindl commented on getReducedProblem.m:
```diff
@@ -0,0 +1,144 @@
+function [newParameters,newProperties,newObjectiveFunction,newOptions] = getReducedProblem(parameters,properties,objectiveFunction,options)
```
Please document this function and those below

paulstapor replied:
General comment on this function:

This function is not used in the current code version. Jan started working on it in #121 to restructure the way how fixed parameters are currently handled.
However, I think this PR fixes enough issues already and we  should not include these files right now, even if we agreed on using them later. I suggest leaving the current implementation for fixed parameters for this PR.
Any objections to this idea?

JanHasenauer replied:
I agree.
(I just want to mention here that most of the functions do not support fixed parameters. This should be addressed soon.)

paulstapor replied:
Important to mention that, yes.
To the best of my knowledge: getMultiStarts and getParameterProfiles should support this feature atm.

JanHasenauer replied:
profiles did not work for me.

JanHasenauer replied:
However, the details do not matter. Let's separate the issue / fixes.

dweindl commented on getReducedProblem.m:
```diff
@@ -0,0 +1,144 @@
+    newParameters.max = parameters.max(freePars);
+    newParameters.guess = parameters.guess(freePars);
+    newParameters.number = length(freePars);
+    if isfield(parameters, 'constraints')
```
Please use some ANDs instead 3 levels of IFs and redundant assignment of defaults. Below: same

dweindl commented on private/barrierFunction.m:
```diff
@@ -0,0 +1,102 @@
+%% Barrier function interface for optimization
+%
+% Last change: Paul Stapor, 06/04/17
```
I think we could let git take care of that...

dweindl commented on private/barrierFunction.m:
```diff
@@ -0,0 +1,102 @@
+%% Documentation of barrierFunction
+%
+% This function applies barrier or penalty funcions to an objective
+% function vvalue and its gradient. The precise method can be chosen by the
```
"vvalue"

output arguments are unclear

dweindl commented on private/coordinateSearch.m:
```diff
@@ -0,0 +1,267 @@
+            end
+        end
+        
+        % apparently bad effects
```
Remove commented-out block?

dweindl commented on private/dynamicHillClimb.m:
```diff
@@ -0,0 +1,391 @@
+    f_output(denormalize(ybst),fbst,funEvals,'iter',outputFcn); % first iteration with start point and jIter = 0
+end
+
+while (~done)
```
100 line loop body, 5 levels of indentation. can't follow that :crying_cat_face:

paulstapor replied:
@yannikschaelte : Could you maybe rearrange this part of the code? I think Daniel is right here, it is kind of hard to read...
You could maybe use something like a struct for the current optimizer state, in which different variables can be grouped. Then it would be easier to transfer code parts to subfunctions, as you could pass maybe 3 instead of 10 arguments...

yannikschaelte replied:
improved readability a bit, added comments, reduced passed parameters. however, it is also good to see which values exactly are passed to subfunctions so as to know what the decision the subfunction makes will depend on.
concerning the length of the main function: i don't think it would help to move parts to subfunctions, it's simply a long and intertwined decision structure.
can try it though if you think it'd be better.

FFroehlich commented on getParameterProfiles.m:
```diff
@@ -69,178 +69,198 @@
+    options.P.min = parameters.min;
+    options.P.max = parameters.max;
+    if isempty(options.profileOptimizationOptions)
+        options.profileOptimizationOptions = options.localOptimizerOptions;
```
👍

FFroehlich commented on getParameterProfiles.m:
```diff
@@ -69,178 +69,198 @@
-        options.profileReoptimizationOptions.display = 'off';
-end
+    end    
+    if isempty(options.parameter_index)
```
please simplify this `if else` clause by setting `options.parameter_index` to `1:parameters.number` and remove code duplication.

FFroehlich commented on getParameterProfiles.m:
```diff
@@ -69,178 +69,198 @@
+            case 'mixed'
+                if (length(unique([options.profile_optim_index, options.profile_integ_index])) ~= length([options.profile_optim_index, options.profile_integ_index])) ...
+                        || (length([options.profile_optim_index, options.profile_integ_index]) ~= length(options.parameter_index))
+                    error('Inconsistent settings for indices in profile calculation.');
```
"Inconsistent" won't help you fixing the error, say that numer of (unique) arguments does not agree.

paulstapor replied:
You're right, the current code is messy...
I changed to whole thing to something more logic...

FFroehlich commented on getParameterProfiles.m:
```diff
@@ -69,178 +69,198 @@
+    end
+    
+    % Check that parameters for which profiles are computed are not fixed
+    if ~isempty(options.profile_optim_index)
```
condense `if` by using `and`

FFroehlich commented on getParameterProfiles.m:
```diff
@@ -69,178 +69,198 @@
+            warning('Fixed parameters are removed from options.profile_optim_index.');
+        end
+    end
+    if ~isempty(options.profile_integ_index)
```
condense `if` by using `and`

FFroehlich commented on getParameterProfiles.m:
```diff
@@ -69,178 +69,198 @@
+            case 'mixed'
+                if strcmp(options.comp_type,'sequential')
+                    for j = options.parameter_index
+                        currentOptions = options.copy();
```
I don't think we have a copy instruction of options anymore.

FFroehlich replied:
also I find "current..." misleading, simply call it "temp..."?

FFroehlich commented on getPropertyProfiles.m:
```diff
@@ -98,15 +98,18 @@
 end
 
 %% Initialization and figure generation
 fh = [];
```
this function (+ variations) seems to appear in multiple functions. Can't we create a seperate function that returns options+fh?

paulstapor replied:
We can and probably should. I opened #154 for this.

FFroehlich commented on getReducedProblem.m:
```diff
@@ -0,0 +1,144 @@
+    end
+
+    % Multi-start optimization results
+    if isfield(parameters,'MS')
```
I would argue that if we do this for `MS`, we should also to this for `S` and `P`.

FFroehlich commented on plotConfidenceIntervals.m:
```diff
@@ -147,15 +148,19 @@
-                    if ((CI(indexSet(iP),2,k)) == inf)
-                        CI(indexSet(iP),2,k) = pStruct.max(indexSet(iP));
+                if (j == 1) && isnan(CI(indexSet(iP),1,1))
+                    % text(pStruct.MS.par(indexSet(iP),1),j,'$\emptyset$','Interpreter','latex','HorizontalAlignment','center','VerticalAlignment','middle')
```
`if(not(...))` and remove commented line?

FFroehlich commented on plotParameterUncertainty.m:
```diff
@@ -251,30 +251,21 @@
+                    ind_not_nan = find(~isnan(diag(parameters.MS.hessian(:,:,1))));
+                    j = find(i==ind_not_nan);
+                    if ~isempty(j)
+                        % Standard deviation of Gaussian approxiamtion of
```
typo

FFroehlich commented on private/direct.m:
```diff
@@ -0,0 +1,750 @@
+function [ret_minval,final_xatmin,history] = direct...
```
can we please give this a more informative name?

yannikschaelte replied:
is the algorithm's name ;) was not planned to be integrated with pesto by now, have deleted the file.

FFroehlich commented on private/getNextProfilePoint.m:
```diff
@@ -109,4 +113,15 @@
+    elseif strcmp(optimizer, 'lsqnonlin')
+        [~, ~, J] = objFun(theta);
+    else
+        error('Unknown optimzer forprofile calculation!');
```
`optimzer` `forprofile`

FFroehlich commented on private/getParProfilesByIntegration.m:
```diff
@@ -752,44 +759,47 @@
-        if (rcond(A2) < options.solver.minCond) || isnan(rcond(A2))
-            dth1 = pinv(A2, options.solver.eps) * b2;
+        if (rcond(A) < options.solver.minCond) || isnan(rcond(A))
+            dth = pinv(A, options.solver.eps) * b;
```
please use more informative variable name than `dth`

FFroehlich commented on private/getParProfilesByIntegration.m:
```diff
@@ -752,44 +759,47 @@
     end
     
     % Check, if parameter bounds are violated
     lRebuild = 0;
```
what does the `l` in `lRebuild` stand for?

paulstapor replied:
for "logical"... seems not to be a good name though, I'll change that

FFroehlich commented on private/getParProfilesByOptimization.m:
```diff
@@ -68,44 +68,47 @@
-        % (fmincon needs the neagtive log posterior for optimization)
-        % with the profile parameter fixed
-        optimizeProfileForParameterI(parameters, objective_function, iPar, options, fh);
+    %% Check for fixed parameters
```
whats the issue here?

paulstapor replied:
The issue was that the current implementation is indeed not working for profiles...

FFroehlich commented on private/getParProfilesByOptimization.m:
```diff
@@ -246,24 +249,44 @@
+                            parameters.min(I),...   % lower bound
+                            parameters.max(I),...   % upper bound
+                            options.profileOptimizationOptions);
+                        [~, ~, J_opt] = negLogPostReduced(theta_I_opt);
```
I am pretty sure we can use the output of `lsqnonlin` to save us this call to `negLogPostReduced`

paulstapor replied:
Well... I'm not so sure about that...
The point is that using lsqnonlin and estimating the standard deviation forces you to add a constant term to the likelihood. This constant term may depend on your implementation (unless we require the user to stick to a specific form).
As Pesto doesn't now this constant term, there's no way of getting it out of lsqnonlin, which would mean that the likelihood values from lsqnonlin and fmincon are not directly comparable...
This call to negLogPostReduced is there for computing the likelihood without this offset...
If anyone has a better idea how to deal with this problem, it would be very welcome.

JanHasenauer replied:
I would probably also remove the call. If I understand correctly, it requires the used to implement two functions. I would rather provide a warning to inform the user that he/she has to handle it outside.

paulstapor replied:
No, the user has to provide only one cost function. It has the following ways of being called:
[residuals] = objectiveFunction(...)
[residuals, Jacobian] = objectiveFunction(...)
[residuals, ~, log-likelihood] = objectiveFunction(...)

JanHasenauer replied:
Ah, okay. Thanks for the clarification.

FFroehlich replied:
Ehm, isn't this really messing things up when you use lsqnonlin for profile calculation? Using the same name for different things (i.e. the same as internal values in optimization vs different) is asking for trouble. If the objective function and the log-likelihood do not match indicate this by using respective variable names.

JanHasenauer replied:
Hm, the use of different names would require changes in many other routines, e.g. visualization routines.

FFroehlich replied:
I would guess that most of these routines currently assume that the objective function == the likelihood function, so they will probably need updates anyways. For the visualization routines this is a matter of copy&paste, so this should not be an issue.

I am primarily worried about functions like `optimizeProfileForParameterI`, where this might easily get mixed up.

paulstapor replied:
I opened up #152 for this.
The point is: The implementation right now works for getMultiStarts and getParameterProfiles. getPropertyProfiles() will be complicated anyway, as this uses nonlinear equality constraints, of which I'm not sure whether this is possible with lsqnonlin...

paulstapor replied:
Any votes whether I should change the implementation for the current PR?

FFroehlich replied:
I am not so much worried about the implementation of the likelihood function. If the function signature is well documented, this should be fine. What I am worried about is the way the values are stored after optimization. 
If something is called J_opt, I assume this is equal to the objective function value that is used in the optimization. If I use J_opt as reference point for my profile calculation and directly use the result from the optimizer, I will be screwed and calculate wrong profiles, as I am missing the offset.
Also, creating different variables for J and nllh will make the code more transparent and reasonable, as it should become evident why we need to call the likelihood function again (you actually don't, you can compute the offset once and add it after every optimization).

paulstapor replied:
It is consistent between profiling and optimization right now. So you can use either lsqnonlin or fmincon for optimization and profiling and switch between those two. The only thing you need to make sure when switching is to provide the correct objective function.

JanHasenauer replied:
I like having always the log-posterior stored somewhere. Converting values manually can be rather cumbersome at some point. Maybe, we could simply include for the use of lsqnonlin two fields, e.g. neglogPost_opt and chi2_opt. To ensure consistency, if residuals are used, both are provided but only neglogPost_opt is used for downstream stuff.

In this case, we should also state neglogPost_opt and chi2_opt instead of J_opt. In the profile calculation routines, we could simply used neglogPost_opt for the stepping, ... by specifying the interface.

paulstapor replied:
Well, right now, negLogPost is stored always. In optimization and profiling.
It would nevertheless make sense to also store chi2 for Lsqnonlin. I pushed a version with a better naming today. However, is there something on which we can agree?
Would everyone be fine with e.g. always storing the log-likelihood? And computing the offset once and using it later on?

FFroehlich commented on private/objectiveWrap.m:
```diff
@@ -18,6 +18,7 @@
+    optimizer = wrapperOptions{5};
     countErrors = wrapperOptions{6};
     showWarnings = wrapperOptions{7};
     if isempty(fixedTheta)
```
correct me if I am wrong but I thought that everything related to options.fixedParameters is taken care of in getReduced...

paulstapor replied:
Not yet:
Jan made up the corresponding files in issue #121 , but so far these functions are not yet used in Pesto. I suggest leaving them out in this PR and addressing issue #121 in another PR.

BBallnus replied:
I agree, this is beyond scope of this pull request

FFroehlich commented on private/performOptimizationBobyqa.m:
```diff
@@ -0,0 +1,51 @@
+function parameters = performOptimizationBobyqa(parameters, negLogPost, iMS, par0, J_0, options)
+
+    % Definition of index set of optimized parameters
+    freePars = setdiff(1:parameters.number, options.fixedParameters);
```
correct me if I am wrong but I thought that everything related to options.fixedParameters is taken care of in getReduced...

FFroehlich commented on private/performOptimizationCs.m:
```diff
@@ -0,0 +1,59 @@
+function parameters = performOptimizationCs(parameters, negLogPost, iMS, par0, J_0, options)
```
what does C stand for in the function name?

yannikschaelte replied:
coordinate search, one might go for the long name here

FFroehlich replied:
yes, please change it here

FFroehlich commented on private/performOptimizationFmincon.m:
```diff
@@ -0,0 +1,67 @@
+function parameters = performOptimizationFmincon(parameters, negLogPost, iMS, par0, J_0, options)
```
if you split this in computation and results assignment, you could reuse the code for profile computation. I suppose doing this until monday is asking for a lot, so please open a corresponding issue.

FFroehlich commented:
General comment: maybe give a bit more informative names to functions. For optimization we have ‘performOptimizationXXX’, for sampling wrappers, the ‘Sampling’ in the middle is missing. Would be nice to also hape optimizer,sampler and profiler prefixes for better identification of what is contained in individual files. Keep in mind that filenames do not have to match function calls (yet it does not hurt if they do as respective doe will be better readable). 

We might also want to add additional subfolders for profiling, optimization and sampling, but I don’t know how well that works in the private folder.

dweindl commented:
I agree, but subfolders are not possible with the /private folder. I never really liked the private folder, but created it to not clutter up the everybody's namespace with all PESTO functions. At that time I didn't know about https://de.mathworks.com/help/matlab/matlab_oop/scoping-classes-with-packages.html. It probably makes sense to introduce some directory tree to split things into /common /optimization /sampling ... but NOT adding the subdirectories to the matlab path.

FFroehlich commented on examples/jakstat_signaling/mainJakstatSignaling.m:
```diff
@@ -50,12 +50,12 @@
-    warning('This PESTO example uses the AMICI toolbox (available at https://github.com/ICB-DCM/AMICI). Unfortunately, there was a problem with AMICI when trying to run this example file. Please check if AMICI is properly installed to run this example. The original error message was:');
-    rethrow(ME);
-end
+% try
```
reenable?

paulstapor replied:
Thanks, that's the kind of stuff that happens while testing... -.-*

FFroehlich commented on examples/rafmekerk_signaling/mainRafMekErkSignaling.m:
```diff
@@ -42,12 +42,12 @@
-    warning('This example uses the additional toolbox AMICI for ODE simulation (freely available at https://github.com/ICB-DCM/AMICI). It seems that AMICI is not or not proeperly installed, since using it resulted in an error. The original error message was:');
-    rethrow(ME);
-end
+% try
```
reenable?

paulstapor commented:
Okay, I hope that by now, all requested changes and remaining issues are addressed. The following is new:
 * I split optimization routines and assignment of results. Results are now assigned in getMultiStarts again and the optimization routines are also used for Profile computation, which reduces duplicated code.
 * I unified the handling of fixed parameters and profiling. They both use the same wrapping now, hence profiling also works for fixed parameters now.
 * I changed the handling of Lsqnonlin to reduce computational overhead (i.e. the offset between log-posterior and squared residuals is only computed once and reused then)
 * I tested all examples, different optimizers, etc.
Please re-check however, if your examples work @LoosC and @BBallnus , because I can not judge if they are okay...

dweindl commented:
Hint: Changing 

> Fixing the following issues: #144 , #143 , #142 , #131 , #124 , #118 , #112 , #108 , #107 , #100

to

> Fixes #144, #143, #142, #131, #124, #118, #112, #108, #107, #100

Will close these issues automatically when merging.
